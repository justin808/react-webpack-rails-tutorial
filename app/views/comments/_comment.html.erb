<!-- TODO: remove? -->
<div class="comment">
  <h2><%= @comment.author %></h2>
  <span><%= @comment.text %></span>
</div>
