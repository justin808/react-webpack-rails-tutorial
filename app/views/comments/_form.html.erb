<%= form_for(comment, html: { class: "flex flex-col gap-4" }) do |f| %>
  <% if comment.errors.any? %>
    <div id="error_explanation" class="prose bg-pink-100 p-4 mb-4 border border-pink-200 rounded text-red-800 prose-strong:text-red-800 prose-ul:my-1">
      <h2><%= pluralize(comment.errors.count, "error") %> prohibited this comment from being saved:</h2>

      <ul>
      <% comment.errors.full_messages.each do |message| %>
        <li><%= message %></li>
      <% end %>
      </ul>
    </div>
  <% end %>

  <div class="field">
    <%= f.label :author, 'Your Name' %><br>
    <%= f.text_field :author, class: "px-3 py-1 leading-4 border border-gray-300 rounded" %>
  </div>
  <div class="field">
    <%= f.label :text, 'Say something using markdown...' %><br>
    <%= f.text_area :text, class: "px-3 py-1 leading-4 border border-gray-300 rounded" %>
  </div>
  <div class="actions">
    <%= f.submit 'Post', class: "self-start px-3 py-1 font-semibold border-0 rounded text-sky-50 bg-sky-600 hover:bg-sky-800 cursor-pointer" %>
  </div>
<% end %>
