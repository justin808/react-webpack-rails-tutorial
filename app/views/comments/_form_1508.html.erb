<%= form_for(@comment, html: { class: "flex flex-col gap-4" }) do |f| %>
  <% if @comment.errors.any? %>
    <div id="error_explanation" class="prose bg-pink-100 p-4 mb-4 border border-pink-200 rounded text-red-800 prose-strong:text-red-800 prose-ul:my-1">
      <h2><%= pluralize(comment.errors.count, "error") %> prohibited this comment from being saved:</h2>

      <ul>
      <% comment.errors.full_messages.each do |message| %>
        <li><%= message %></li>
      <% end %>
      </ul>
    </div>
  <% end %>

  <div class="field">
    <%= f.label :author, 'Your Name' %><br>
    <%= f.text_field :author, class: "px-3 py-1 leading-4 border border-gray-300 rounded" %>
  </div>
  <div class="field">
    <%= f.label :text, 'Say something using markdown...' %><br>
    <%= f.text_area :text, class: "px-3 py-1 leading-4 border border-gray-300 rounded" %>
  </div>
  <div class="actions">
    <%= f.submit 'Post', class: "self-start px-3 py-1 font-semibold border-0 rounded text-sky-50 bg-sky-600 hover:bg-sky-800 cursor-pointer" %>
  </div>

  <h1>Below is a react component which render inside a form (that will proof #1508 is fixed), You can turn-off the flag `force_load` and this below form will not show. </h1>
  <div class="mx-10 border border-red-600 rounded">
    <%= react_component('HotwiredCommentForm', props: { }, prerender: false, force_load: true) %>
  </div>
<% end %>
