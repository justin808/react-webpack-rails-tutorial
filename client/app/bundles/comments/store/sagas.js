import { takeEvery } from 'redux-saga';
import { call, put, fork } from 'redux-saga/effects';
import requestsManager from 'libs/requestsManager';
import * as actionTypes from '../constants/commentsConstants';
import * as actionCreators from '../actions/commentsActionCreators';

function* fetchComments() {
  try {
    const responce = yield call(requestsManager.fetchEntities);
    yield put(actionCreators.fetchCommentsSuccess(responce.data));
  } catch (error) {
    yield put(actionCreators.fetchCommentsFailure(error));
  }
}

function* submitComment(action) {
  try {
    const responce = yield call(requestsManager.submitEntity, action);
    yield put(actionCreators.submitCommentSuccess(responce.data));
  } catch (error) {
    yield put(actionCreators.submitCommentFailure(error));
  }
}

export function* fetchCommentsSaga() {
  yield* takeEvery(actionTypes.SET_IS_FETCHING, fetchComments);
}

export function* submitCommentSaga() {
  yield* takeEvery(actionTypes.SET_IS_SAVING, submitComment);
}

export default function* root() {
  yield fork(fetchCommentsSaga);
  yield fork(submitCommentSaga);
}
