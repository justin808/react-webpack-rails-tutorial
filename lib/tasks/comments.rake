desc "Remove all comments older than one day"
task :rmcom => :environment do
  puts "Removing comments older than one day"
  Comment.delete_all(:created_at => Time.now.midnight - 1.year..Time.now.midnight - 1.day)
end
