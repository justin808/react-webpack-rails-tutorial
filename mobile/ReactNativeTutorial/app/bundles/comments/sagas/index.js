import { Alert } from 'react-native';
import { takeLatest } from 'redux-saga';
import { call, put, select } from 'redux-saga/effects';
import { Actions as navigationActions } from 'react-native-router-flux';
import { reduxUtils } from '../../../utils';
import commentsStoreSelector from '../../../selectors/commentsStoreSelector';
import commentFormSelector from '../../../selectors/commentFormSelector';

import { actions as reduxActions } from '../../../reducers';
import * as api from '../../../api';

const FETCH = 'SAGA:COMMENTS:FETCH';
const CREATE_COMMENT = 'SAGA:COMMENTS:CREATE';
const UPDATE_FORM = 'SAGA:COMMENTS:UPDATE_FORM';


function* fetchHandler() {
  yield put(reduxActions.setLoadingComments(true));
  try {
    const response = yield call(api.fetchComments);
    yield put(reduxActions.createComments(response.entities.comments));
  } catch (e) {
    console.log(e);
    yield call(Alert.alert, 'Error', 'Could not connect to server', [{ text: 'OK' }]);
  } finally {
    yield put(reduxActions.setLoadingComments(false));
  }
}

function* fetchSaga() {
  yield* takeLatest(FETCH, fetchHandler);
}


function* updateFormHandler(action) {
  yield put(reduxActions.updateCommentForm(action.payload));
}

function* updateFormSaga() {
  yield* takeLatest(UPDATE_FORM, updateFormHandler);
}


function* createCommentHandler() {
  const state = yield select();
  const commentsStore = commentsStoreSelector(state);
  const tempId = reduxUtils.getNewId(commentsStore);

  const comment = commentFormSelector(state).merge({ id: tempId }).delete('meta');
  const reduxComment = { [tempId]: comment.toJS() };
  yield put(reduxActions.createComments(reduxComment));
  yield call(navigationActions.pop);
  try {
    const response = yield call(api.postComment, comment);
    yield put(reduxActions.createComments(response.entities.comments));
    yield put(reduxActions.resetCommentForm());
  } catch (e) {
    console.log(e);
    yield call(Alert.alert, 'Error', 'Could not post your comment', [{ text: 'OK' }]);
  } finally {
    yield put(reduxActions.removeComment(tempId));
  }
}

function* createCommentSaga() {
  yield* takeLatest(CREATE_COMMENT, createCommentHandler);
}

export default [
  fetchSaga,
  updateFormSaga,
  createCommentSaga,
];

const fetch = () => ({ type: FETCH });
const updateForm = (payload) => ({ type: UPDATE_FORM, payload });
const createComment = (payload) => ({ type: CREATE_COMMENT, payload });

export const actions = {
  fetch,
  updateForm,
  createComment,
};
