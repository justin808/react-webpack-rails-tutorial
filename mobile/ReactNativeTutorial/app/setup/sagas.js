import _ from 'lodash/fp';

import commentsSagas from '../bundles/comments/sagas';

const sagas = [
  ...commentsSagas,
];

const sagaIterators = _.map(saga => saga && saga(), sagas);

function* appSaga() {
  yield sagaIterators;
}

export default appSaga;
