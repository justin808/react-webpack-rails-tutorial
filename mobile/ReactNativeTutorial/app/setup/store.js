import { applyMiddleware, compose, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';

import loggerMiddleware from './loggerMiddleware';
import appReducer, { initialState } from '../reducers';
import appSaga from './sagas';

const sagaMiddleware = createSagaMiddleware();

let composedStore;
if (__DEV__) {
  composedStore = compose(applyMiddleware(sagaMiddleware, loggerMiddleware));
} else {
  composedStore = compose(applyMiddleware(sagaMiddleware));
}
const store = composedStore(createStore)(appReducer, initialState);

sagaMiddleware.run(appSaga);

export default store;
