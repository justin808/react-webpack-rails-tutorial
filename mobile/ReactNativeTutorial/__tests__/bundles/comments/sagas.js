import store from 'redux-mock-store';

import { actions } from 'ReactNativeTutorial/app/bundles/comments/sagas';

describe('fetch', () => {
  it('fetches comments from remote api', () => {
    const response = {
      comments: [
        { id: 1, author: 'Alexey', text: 'Just some random comment' },
        { id: 2, author: 'Justin', text: 'Another random comment' },
      ],
    };
    fetch.mockResponseSuccess(response);
  });
});
