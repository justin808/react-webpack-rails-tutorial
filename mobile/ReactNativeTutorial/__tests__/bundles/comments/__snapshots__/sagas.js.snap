exports[`fetch fetches comments from remote api 1`] = `
Array [
  Object {
    "type": "SAGA:COMMENTS:FETCH",
  },
  Object {
    "loading": true,
    "type": "COMMENTS_STORE:SET_LOADING",
  },
]
`;
